text/0000-release-meta-spec-v2.md

```diff
@@ -0,0 +1,305 @@
+3.  Fetch the release `META.json` file, read in the `releases/pgxn` object,
+    and use PGXN's public key to verify that it was signed by PGXN.
+4.  Decode the payload and use its `uri` field to download the release zip
+    file.
+5.  Compare the the `sha512` digest from the payload to a digest generated
+    from the downloaded zip file.
+6.  If they digests are the same, continue with building and installing.
+    Otherwise abort with an error.
+
+## Reference-level explanation
+
+This design ensures proper validation of PGXN releases by the following chain
+of trust:
+
+1.  A root key pair is maintained by PGXN, with the private key kept offline.
+2.  A release key pair is generated and signed by the private root key, with
+    the private key kept in an online vault accessible only to PGXN Manager.
+3.  The public keys for both keys are published by PGXN.
+4.  PGXN Manager uses the private release key to sign releases as described
+    above. The most important property in the signed payload is the list of
+    digests.
+5.  Clients can verify the signature with the public release and root keys.
+6.  With the data validated, the client can download and verify the release
+    file against a signed digest. In this manner, the authenticity of the
+    release file can be verified all the way to the root key.
+
+To support this infrastructure, PGXN Manager **MUST** be updated to properly
+generate and sign the payload and include it in the release `META.json` files.
+Clients **MUST** follow the [JWS validation steps].
+
+## Drawbacks
+
+*   This pattern could make it more difficult for clients to install code from
+    PGXN, especially if they incorrectly validate the signature.
+*   Some clients may choose not to implement the validation, potentially
+    leaving users to think they have trusted, validated code when they may
+    not.
+
+## Rationale and alternatives
+
+*   This design takes advantage of the [JWS] standard to naturally augment the
+    PGXN `META.json` format to enable best-in-class digital signatures and
+    validation.
+*   Without key signing, questions will gradually mount as to the
+    trustworthiness of releases published on PGXN. In the era of [what's teh
+    word I want here?] compromises, it's essential for PGXN to enable
+    compromise detection.
+
+## Prior art
+
+This design was inspired by the [Python wheel] format, particularly its
+precedent of signing only the hash digests for a package rather than the
+package itself. The use of [JWS JSON Serialization] varies from [Python
+wheel], and is enabled by the separation of the release `META.json` file from
+the release file it describes.
+
+The use of [JWS] ensures a widely-vetted key signing and distribution
+standard, and the likelihood that clients can take advantage of well-tested,
+mature libraries to handle signing and validation. And finally, its design
+allows for key rotation when necessary.
+
+## Unresolved questions
+
+None currently.
```

Comment:
If public keys are distributed through the same distribution method as the packages (i.e., PGXN web services) what prevents the whole system from being compromised?  Seems like a different distribution method (at least from a different host) would alleviate at least some of this concern.

Reply:
Also in order to augment the `META.json` file from the distribution's copy, we're be default changing the hash of the `META.json` file iteself—does there need to be some sort of validation that the only changes to that file have come from the added keys themselves?

Reply:
This RFC doesn't address key management; I'll add that to the unresolved questions.

The file that includes the signature is not included in the distribution file; it's separate. The `META.json` included in the zip file is unchanged from how authors provide it (IIRC), and is defined by #3.

This RFC _extends_ on #3 by adding this additional metadata, and the resulting "release metadata" file is distributed along side the zip file, not in it. This pattern is true for the original PGXN design, too, where each release results in two files:

* `dist/$name/$version/META.json` --- the release metadata file
* `dist/$name/$version/$name-$version.zip` --- includes the distribution metadata file

I tried to describe that process in the `Publishing a Release` section.

Reply:
In fa66f7a, I added a bunch of unresolved questions, including about key management. I also went into a bit more detail on these files and how they're not included in the source distribution file.